### PR TITLE
weight initialize

### PR DESCRIPTION
**Pull request by r08922a10**
* Weight initialization method : init_weight in class Agent
* New attribute in class Simulation : baseline_agent

### CODE REVIEW

File: rl_reinforce.py [hunk 1]
```diff
@@ -13,7 +13,7 @@
 import matplotlib.pyplot as plt
 
 np.random.seed(666)
-torch.manual_seed(666)
+#torch.manual_seed(666)
```
Comment (r08922a10) [inline rl_reinforce.py:16]: 條餐的時候先 random 看穩定性，避免 gpu/cpu 的bias

File: rl_reinforce.py [hunk 2]
```diff
@@ -1136,16 +1156,17 @@ def actor_critic_policy_loss(self):
 
         accumulated_rewards_immediate = torch.tensor(accumulated_rewards_immediate, device=device)
 
-        accumulated_rewards_terminal = (accumulated_rewards_terminal - accumulated_rewards_terminal.mean()) / (accumulated_rewards_terminal.std() + 1e-9)
+        #accumulated_rewards_terminal = (accumulated_rewards_terminal - accumulated_rewards_terminal.mean()) / (accumulated_rewards_terminal.std() + 1e-9)
         
-        accumulated_rewards_immediate = (accumulated_rewards_immediate - accumulated_rewards_immediate.mean()) / (accumulated_rewards_immediate.std() + 1e-9)
+        #accumulated_rewards_immediate = (accumulated_rewards_immediate - accumulated_rewards_immediate.mean()) / (accumulated_rewards_immediate.std() + 1e-9)
```
Comment (r08922a10) [inline rl_reinforce.py:1161]: 由於有masking，這邊的 normalize 會失準，先不要用，除非你們有人用了他會穩定比較好

File: rl_reinforce.py [hunk 3]
```diff
@@ -1374,18 +1395,25 @@ def main():
 
     agent = ActorCriticAgent(**args_agent).to(device)
 
+    baseline_agent = ActorCriticAgent(**args_agent).to(device)
+
+    agent.init_weight(torch.nn.init.normal_, mean=0., std=0.01)
```
Comment (r08922a10) [inline rl_reinforce.py:1400]: 這邊可以條看看，std 會決定他接不接不接近 uniform 的 action 分佈。

File: rl_reinforce.py [hunk 4]
```diff
@@ -1374,18 +1395,25 @@ def main():
...
+    optimizer = optim.Adam(agent.parameters(), lr=5e-2)
 
-    game = Simulatoin(agent, env, optimizer, verbose=True, gamma=0.9)
+    game = Simulatoin(agent, baseline_agent, env, optimizer, verbose=True, gamma=0.9)
 
     game.is_actor_critic = True
 
-    r_0, s_0, a_0 = game.testing(max_steps=300, max_episode=50, greedy=False)
+    #r_0, s_0, a_0 = game.testing(max_steps=300, max_episode=50, greedy=False, baseline=True) # Random walk
+    r_0, s_0, a_0 = game.testing(max_steps=300, max_episode=50, greedy=True, baseline=True) # No action
```
Comment (r08922a10) [inline rl_reinforce.py:1413]: comment 掉的那一行是 random walk，而底下的 greedy 會是 no action，原因是 argmax 會先 sort 過，而理論上 pytorch 或是 numpy 的 argsort default sort 是 increase 排列，所以會取最後一個，就是 no action。如果有問題再跟我說。

**Review by r08922a27 — APPROVED** [review]

**Review by r08922196 — APPROVED** [review]
